### PR TITLE
Added Spring Security configuration and Role entity for user authoriz…

### PR DESCRIPTION
**Pull request by VKhmil**
…ation

Implemented Spring Security configuration to enable Basic authentication and added Role entity to define user roles. Registered users are now assigned roles, allowing access to specific endpoints based on their role. Updated Liquibase scripts to insert roles into the database and assign roles to existing users, ensuring data integrity and consistency. Security constraints are applied to endpoints based on user roles. Swagger UI is now accessible after configuring security filters.

### CODE REVIEW

File: src/main/java/com/bookappstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity httpSecurity) throws Exception {
+        return httpSecurity.cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/api/auth/***")
```
Comment (Rommelua) [inline src/main/java/com/bookappstore/config/SecurityConfig.java]: image

File: src/main/java/com/bookappstore/repository/user/UserRepository.java [hunk 2]
```diff
@@ -3,7 +3,12 @@
 import com.bookappstore.model.User;
 import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
 public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("FROM User u LEFT JOIN FETCH u.roles r "
+            + "WHERE u.email = :email "
+            + "AND u.isDeleted = FALSE "
+            + "AND r.isDeleted = FALSE")
```
Comment (Rommelua) [inline src/main/java/com/bookappstore/repository/user/UserRepository.java]: What is the reason for using `@Query` here? If for JOIN FETCH only - better use `@EntityGraph`

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/bookappstore/repository/user/UserRepository.java (hunk 2 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/repository/user/UserRepository.java]: Nice tip, thanks

File: src/main/java/com/bookappstore/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@EnableMethodSecurity
+@EnableWebSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/config/SecurityConfig.java:15]: Do you really need two annotations here?

File: src/main/java/com/bookappstore/controller/BookController.java [hunk 4]
```diff
@@ -51,14 +55,24 @@ public List<BookDto> search(BookSearchParametersDto bookSearchParametersDto) {
 
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/controller/BookController.java]: Be consistent, why in some places you use "hasRole('ROLE_ADMIN')" and in other "hasRole('ADMIN')")
Also, think about which endpoint could use USER and ADMIN?

File: src/main/java/com/bookappstore/model/User.java [hunk 5]
```diff
@@ -30,6 +40,48 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id"))
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/User.java]: It's better to initialise when you declare
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/com/bookappstore/model/Role.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE roles SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "roles")
+public class Role {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/Role.java]: You can implement GrantedAuthority and override 
```
public String getAuthority() {
        return name.name();
    }
```

File: src/main/java/com/bookappstore/model/User.java [hunk 7]
```diff
@@ -30,6 +40,48 @@ public class User {
...
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return roles.stream()
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/User.java]:
```suggestion
        return roles;
```

File: src/main/java/com/bookappstore/model/Role.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "name", nullable = false, unique = true)
+    private RoleName name;
+    @Column(name = "is_deleted", nullable = false)
+    private boolean isDeleted = false;
+
+    public enum RoleName {
+        ADMIN,
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/model/Role.java:39]:
```suggestion
        ROLE_ADMIN,
        ROLE_USER
```

File: src/main/java/com/bookappstore/repository/user/UserRepository.java [hunk 9]
```diff
@@ -2,8 +2,11 @@
 
 import com.bookappstore.model.User;
 import java.util.Optional;
+import org.springframework.data.jpa.repository.EntityGraph;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface UserRepository extends JpaRepository<User, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/repository/user/UserRepository.java]:
```suggestion
```

File: src/main/java/com/bookappstore/repository/user/role/RoleRepository.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package com.bookappstore.repository.user.role;
+
+import com.bookappstore.model.Role;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/repository/user/role/RoleRepository.java]:
```suggestion
```

File: src/main/java/com/bookappstore/security/CustomUserDetailService.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email).orElseThrow(()
+                -> new UsernameNotFoundException("Can't find user by email"));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/security/CustomUserDetailService.java]:
```suggestion
                -> new UsernameNotFoundException("Can't find user by email: " + email));
```

File: src/main/java/com/bookappstore/service/impl/BookServiceImpl.java [hunk 12]
```diff
@@ -56,4 +56,20 @@ public List<BookDto> search(BookSearchParametersDto params) {
                 .map(bookMapper::toDto)
                 .toList();
     }
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto requestDto) {
+        Book bookById = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find by book id " + id)
+        );
+        bookById.setTitle(requestDto.getTitle());
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/BookServiceImpl.java]: It looks like mapper is needed here

File: src/main/java/com/bookappstore/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -23,7 +30,13 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
             throw new RegistrationException("Unable to complete registration");
         }
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
 
+        Role userRole = roleRepository.findByName(Role.RoleName.USER)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/UserServiceImpl.java:35]: You can use 
```
private Role userRole;

@PostConstruct
    public void init() {
        userRole = roleRepo.findByName(Role.RoleName.ROLE_USER);
    }
```

File: src/main/java/com/bookappstore/service/impl/UserServiceImpl.java [hunk 14]
```diff
@@ -23,7 +30,13 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
+                .orElseThrow(() -> new RegistrationException("Can't find role by name"));
+        Set<Role> defaultUserRoleSet = new HashSet<>();
+        defaultUserRoleSet.add(userRole);
+        user.setRoles(defaultUserRoleSet);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookappstore/service/impl/UserServiceImpl.java:39]:
```suggestion
        user.setRoles(Set.of(userRole));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/bookappstore/config/SecurityConfig.java (hunk 3 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/config/SecurityConfig.java:15]: thanks, good tip

File: src/main/java/com/bookappstore/controller/BookController.java (hunk 4 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/controller/BookController.java]: Yes, really my bad

File: src/main/java/com/bookappstore/model/User.java (hunk 5 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/model/User.java]: good tip, my bad

File: src/main/java/com/bookappstore/model/Role.java (hunk 6 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/model/Role.java]: Nice, much better

File: src/main/java/com/bookappstore/service/impl/BookServiceImpl.java (hunk 12 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/service/impl/BookServiceImpl.java]: Thanks, nice tip

File: src/main/java/com/bookappstore/service/impl/UserServiceImpl.java (hunk 13 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/service/impl/UserServiceImpl.java:35]: nice tip

File: src/main/java/com/bookappstore/service/impl/UserServiceImpl.java (hunk 14 above)
Comment (VKhmil) [reply src/main/java/com/bookappstore/service/impl/UserServiceImpl.java:39]: nice tip

**Pull request merged**